identity_storage/src/key_storage/key_storage_trait.rs

```diff
@@ -0,0 +1,62 @@
+// Copyright 2020-2023 IOTA Stiftung
+// SPDX-License-Identifier: Apache-2.0
+
+pub use crate::key_storage::KeyStorageError;
+pub use crate::key_storage::KeyStorageErrorKind;
+use crate::key_storage::SignatureAlgorithm;
+
+use crate::key_storage::KeyId;
+use crate::key_storage::KeyType;
+use async_trait::async_trait;
+use identity_jose::jwk::Jwk;
+
+use super::key_gen::JwkGenOutput;
+
+pub type KeyStorageResult<T> = Result<T, KeyStorageError>;
+
+#[cfg(not(feature = "send-sync-storage"))]
+mod storage_sub_trait {
+  pub trait StorageSendSyncMaybe {}
+  impl<S: super::KeyStorage> StorageSendSyncMaybe for S {}
+}
+
+#[cfg(feature = "send-sync-storage")]
+mod storage_sub_trait {
+  pub trait StorageSendSyncMaybe: Send + Sync {}
+  impl<S: Send + Sync + super::KeyStorage> StorageSendSyncMaybe for S {}
+}
+
+#[cfg_attr(not(feature = "send-sync-storage"), async_trait(?Send))]
+#[cfg_attr(feature = "send-sync-storage", async_trait)]
+pub trait KeyStorage: storage_sub_trait::StorageSendSyncMaybe {
+  /// Generate a new key represented as a JSON Web Key.
+  ///
+  /// It's recommend that the implementer exposes constants for the supported [`KeyType`].
+  async fn generate_jwk(&self, key_type: KeyType) -> KeyStorageResult<JwkGenOutput>;
+
+  /// Insert an existing JSON Web Key into the storage.
+  ///
+  /// All private key components of the `jwk` must be set.
+  async fn insert_jwk(&self, jwk: Jwk) -> KeyStorageResult<KeyId>;
+
+  /// Sign the provided `data` using the private key identified by `key_id` with the specified `algorithm`.
+  ///
+  /// It's recommend that the implementer exposes constants for the supported [`SignatureAlgorithm`].
+  async fn sign(&self, key_id: &KeyId, algorithm: SignatureAlgorithm, data: Vec<u8>) -> KeyStorageResult<Vec<u8>>;
```

Comment:
That is indeed the crucial question. I can think of two ways we might do this.

## Option 1

The first is the obvious one, where we expose both key type and signature algorithm type to the user and they have to pass in the correct one.

```rust
fn main() {
  let document = CoreDocument::new(...);
  let storage = Storage::new(key_storage, identity_storage);

  document.generate_jwk(&storage, storage_impl::ED25519_KEY_TYPE, "#sign-0");

  let credential = ...

  let proof = document.sign_jwt(credential, storage_impl::ED25519_SIGNING_ALG, "#sign-0");
}
```

- Pro: Lots of flexibility, i.e. a JWK can be used with multiple different signature algorithms, there isn't a forced 1:1 mapping.
- Con: Harder for the user to use correctly.

## Option 2.1

The storage translates from a JWK key type to a signature algorithm, so the user doesn't have to pass it in.

```rust
fn main() {
  let document = CoreDocument::new(...);
  let storage = Storage::new(key_storage, identity_storage);

  document.generate_jwk(&storage, storage_impl::ED25519_KEY_TYPE, "#sign-0");

  let credential = ...

  // At this stage, `CoreDocumentExt` would call `key_storage.map_jwk(jwk, key_type)`
  // to translate the jwk to the signature algorithm type.
  let proof = document.sign_jwt(credential, "#sign-0");
}
```

## Option 2.2

Very similar idea, just using the `JwsAlgorithm` type. We either ask users to supply the `alg` for the JWK when creating it or let the storage decide, thereby creating an association to the (single) algorithm the key can be used with. When calling sign on the storage, the storage can just look at the `alg` in the JWK and know what signature algorithm to use. No user input required at signing-time.

```rust
fn main() {
  let document = CoreDocument::new(...);
  let storage = Storage::new(key_storage, identity_storage);

  // User supplies the algorithm they want to use the key with.
  // Alternatively we could omit this parameter and let the storage decide.
  document.generate_jwk(&storage, storage_impl::ED25519_KEY_TYPE, JwsAlgorithm::EdDSA, "#sign-0");

  let credential = ...

  let proof = document.sign_jwt(credential, "#sign-0");
}
```

For Option 2
- Pro: Easier for the user as they don't have to match the JWK type to the signature algorithm type of the storage.
- Pro: We wouldn't even need `SignatureAlgorithm`.
- Con: `JwsAlgorithm` is not extensible, so we have to keep it up-to-date or users cannot pass in their own variant. Not sure how problematic this really is in practice, probably not very much.
- Con: A key cannot be used with different algorithms. Again, probably not very problematic. Just use a different key. Although the spec says:
> The "alg" (algorithm) parameter identifies the algorithm intended for use with the key.  The values used should either be registered in the IANA "JSON Web Signature and Encryption Algorithms" registry established by [[JWA](https://www.rfc-editor.org/rfc/rfc7517#ref-JWA)] or be a value that contains a Collision-Resistant Name. The "alg" value is a case-sensitive ASCII string. Use of this member is OPTIONAL.

So we could also take either a string or a `JwsAlgorithm` if we wanted to be super flexible.

Reply:
Thanks for writing the proposals and also giving examples in code of what they might look like. 

I think I like `Option 2.2` the most, mainly because of its simplicity. Since not all parameters can be set in the `JwsAlgorithm` (like for example how many bits the key consists of) key storages will have to specify that in their docs, but I think that is fine.  

With regards to extensibility, I think the values present in `JwsAlgorithm` already gives a decent amount of flexibility. As we've mentioned before sometimes less flexibility results in more interoperability, so not always necessarily a bad thing. More (advanced) cryptographic algorithms could perhaps be supported in the future using another kind of `KeyStorage` that is dedicated to work with the upcoming `DataIntegrity` W3C specification.

Reply:
Nice, I think I also like 2.2 best. I'll try to make it a thing.

The key size for example can be set in the key type, e.g.

```rust
pub const RSA_2048_KEY_TYPE: KeyType = KeyType::from_static_str("RSA_2048);
pub const RSA_4096_KEY_TYPE: KeyType = KeyType::from_static_str("RSA_4096);
// ...
```

I agree that DI might be required for the very advanced use cases where JWT is insufficient.